File: DesignPatterns/01_GOF23/DP_examples/Go/03_Behavioral/Iterator/iterator.go
package Iterator

// Iterator 迭代器接口
type Iterator[T any] interface {
	First()
	Next()
	IsDone() bool
	CurrentItem() T
}

// Iterable 可迭代接口
type Iterable[T any] interface {
	GetIterator() Iterator[T]
}

// IAggregate 聚合接口
type IAggregate[T any] interface {
	Iterable[T]
	Count() int
	Add(item T)
	Remove(item T) bool
	Contains(item T) bool
	Clear()
}

File: DesignPatterns/01_GOF23/DP_examples/Rust/03_Behavioral/Iterator/iterator.rs
// 迭代器接口
pub trait Iterator<T> {
    fn first(&mut self);
    fn next(&mut self);
    fn is_done(&self) -> bool;
    fn current_item(&self) -> Option<T>;
}

// 可迭代接口
pub trait Iterable<T> {
    fn get_iterator(&self) -> Box<dyn Iterator<T> + '_>;
}

// 聚合接口
pub trait IAggregate<T>: Iterable<T> {
    fn count(&self) -> usize;
    fn add(&mut self, item: T);
    fn remove(&mut self, item: &T) -> bool;
    fn contains(&self, item: &T) -> bool;
    fn clear(&mut self);
}
